waspls/src/Wasp/Backend/ConcreteParser/Internal.hs

```diff
@@ -0,0 +1,716 @@
+  -- "Succeed" is identity of <>. The second pattern does affect the result tree,
+  -- but only by changing where trivia tokens are placed in the tree, which is
+  -- not an issue.
+  Succeed <> rule = rule
+  rule <> Succeed = rule
+  first <> second = Chain first second
+
+-- | @mempty@ is a parser that matches an empty string, or a string of only
+-- trivia tokens.
+instance Monoid GrammarRule where
+  mempty = Succeed
+
+-- | Run a "GrammarRule" on an input of tokens. Returns a tuple @(errors, nodes)@.
+parse :: [Token] -> GrammarRule -> ([ParseError], [SyntaxNode])
+parse tokens grammar =
+  let initialState =
+        ParseState
+          { pstateInput = tokens,
+            pstateCurrentLevelNodes = [],
+            pstateErrors = [],
+            pstateNextTokenOffset = 0,
+            pstatePrevTokenOffset = 0,
+            pstateFollowingTokens = []
+          }
+   in case runState (runExceptT (parseRule grammar)) initialState of
+        (Left err, state) -> case err of
+          Unwind _ -> error "Unwind at top-level GrammarRule (impossible, this is a bug in Wasp.Backend.ConcreteParser.Internal)"
+          ParseError perr ->
+            let errs = pstateErrors state ++ [perr]
+                newNode = errorNode perr False
+                nodes = pstateCurrentLevelNodes state ++ [newNode]
+             in (errs, nodes)
+        (_, state) -> (pstateErrors state, pstateCurrentLevelNodes state)
+
+data ParseState = ParseState
+  { -- | Remaining tokens in input.
+    pstateInput :: [Token],
+    -- | "SyntaxNode"s processed so far at the current level of the CST. An
+    -- initial top-level is created when the parser is started and each "Group"
+    -- rule creates a new level.
+    pstateCurrentLevelNodes :: [SyntaxNode],
+    -- | Errors collected that will be reported.
+    pstateErrors :: [ParseError],
+    -- | 0-based offset from the beginning of the source file, in characters,
+    -- of the start of the next token (@head . `pstateInput`@).
+    pstateNextTokenOffset :: !Int,
+    -- | 0-based offset from the beginning of the source file, in characters,
+    -- of the start of the last consumed token
+    pstatePrevTokenOffset :: !Int,
+    -- | Stack of possible first tokens, collected when we enter a group as the
+    -- first grammar rule in a "Chain".
+    --
+    -- A member @None@ represents EOF.
+    pstateFollowingTokens :: [TokenSet]
+  }
+  deriving (Eq, Show, Ord)
+
+data ParseException
+  = -- | @Unwind tokenKind@ is used as a signal during error recovery. It is used
+    -- to implement the unwinding behavior described at the top of the module.
+    --
+    -- @tokenKind@ is the kind of the token that an error was produced on.
+    Unwind (Maybe TokenKind)
+  | -- | @ParseError parseError@ is used as a signal when an unexpected token is
+    -- encountered. This kind of exception is caught by error recovery points
+    -- and added to the output list of errors.
+    ParseError ParseError
```

Comment:
bump

Reply:
I'm not 100% sure what your final suggestion was, but these are the two I see:

1.  Rename `Unwind` to `RecoverableParseError`: Already discussed in another comment, staying with `Unwind` to keep consistency with naming.

2. Get rid of `Unwind` and just re-throw `ParseError`: I agree that performance is probably not a worry. But because we only want to log the error and make an error node once, we would need to add a field to the `ParseError` constructor to keep track of whether its been logged. Also, it feels kind of strange to re-use the same exception type for both "I came across an unexpected thing while parsing" and "I know this next token can be parsed by someone, so pass this message up the tree until we find that someone". So I'd say we should keep both.